### ISSUE TITLE
default keybindings don't always make sense on macOS

### ISSUE THREAD

**Issue by kjyv** [zim-desktop-wiki/zim-desktop-wiki#878]
While some keybindings are properly translated like Ctrl-c into cmd-c etc., navigating text with the keyboard is impossibly hard because cmd-left/right does not work at all (go to beginning/end of line) and alt-left/right is mapped to prev/next page instead of prev/next word. 
As accelerators can not be changed anymore which I did with the gtk2 version, I'm opening this issue. But really it should have been looked at for gtk2 already so it works properly by default. Is there code in zim with specific keybindings for macOS or are the few done by gtk automatically?

**Comment by kjyv** [comment]
I've just seen the new key bindings editor and that works well to disable alt-left/alt-right and add cmd-, for preferences. However, I don't see a way to add cmd-left/right for beginning/end of line which is standard behavior on macOS.

**Comment by arigit** [comment]
@kjyv how did you fix the Alt-left/Alt-right behavior on GTK2 & 3? in other words what did you remap the Alt-left etc keystroke to? I'm trying to get Alt right/left move one word forward/backward but I can't find the actual keywords I need to use for that remapping.

For beginning/end of the line I assume you mapped Ctrl right/left to End/Home

**Comment by kjyv** [comment]
This ticket is about zim >0.70 using GTK3. However, I think in both cases unmapping alt-left/right is necessary to get the proper behaviour.
Also, you can't map ctrl-right/left to other keys, only to a "gtk_accel_path" (I think).
However, ctrl-l/r doesn't go to beginning/end of line on macOS by default anyway, they change workspace. I think you mean cmd-left/right, which I think Gtk is not handling properly.

**Comment by arigit** [comment]
Sorry :) I realize I totally hijacked the Issue discussion

Can I still ask you (since github doesn't allow messaging), how did you solve the "previous/next word shortcut does not exist by default" problem in the old GTK2 Zim on Mac OS? I can't seem to find a way to work around it, so facing the same situation you are facing with GTK3. I am even looking into the source code (since I run zim from the source tree) but no luck thus far

p.s Yes I meant Ctrl and Alt combinations, github stripped the ```<alt> and <cmd>``` as I forgot the quotes.

**Comment by kjyv** [comment]
As written, unmapping the keys should be enough. Gtk 2.0 also does the cmd-left/right part correctly. If that does not work for you, it might be because of how your Gtk is built, not sure.

**Comment by arigit** [comment]
Thanks. I could fix the problem as follows (for anyone with this problem landing here):
1- in Preferences > Keyboard > Shortcuts > Mission Control, I changed the bindings for "Move left/right a space" to Ctrl+Cmd+left/right; that freed up Ctrl+left/right which immediately started to work in Zim for "word navigation". 
2- I used BTT (bettertouchtool) to create a conditional mapping (application = Zim, window= contains "Notebook" to map Alt+left/right into Ctrl+left/right.

This allows me to use Alt+left/right as normal in MacOS. For consistency with Firefox, I also mapped with BTT Ctrl+PageUp/down into the original Alt+left/right (for "tab" navigation in Zim)

**Comment by kjyv** [comment]
Upstream bug for cmd-left/-right: https://gitlab.gnome.org/GNOME/gtk/issues/1958

**Comment by reagle** [comment]
@kjyv what new keybinding editor? I'm using zim 0.71.1 on MacOS.

**Comment by kjyv** [comment]
not yet released, it's in the develop branch

**Comment by reagle** [comment]
@kjyv I looked through the commits in zim's dev branch and don't see anything like this.

**Comment by kjyv** [comment]
Ah, sorry, it's even already in 0.71.1. 
https://github.com/zim-desktop-wiki/zim-desktop-wiki/commit/92d433ecf059d262e289020b2a0631f469e6fd52

**Comment by reagle** [comment]
@kjyv how do you access the editor? I don't see it in preferences or anywhere else.

**Comment by kjyv** [comment]
It is in preferences. Anyway, can we please stop using this ticket as a chat? It is about adding better default keybindings on macOS.

**Comment by michaelwilleredb** [comment]
FYI: CTRL-3 is used on UK keyboards to produce '#' character. This key is silently ignored by the editor. I have checked the keybindings, and nothing is there for that key.

**Comment by michaelwiller** [comment]
FYI: CTRL-3 is used on UK keyboards to produce '#' character. This key is silently ignored by the editor. I have checked the keybindings, and nothing is there for that key.

**Comment by dehesselle** [comment]
I can remove the default binding from ⌥← and ⌥→ to restore macOS default behavior. Should we consider another shortcut as a replacement, like adding the shift key to the mix (e.g.  ⇧⌥← and ⇧⌥→)?

**Comment by kjyv** [comment]
I think there no extra work necessary, ⌥← and ⌥→ work fine in the latest versions for me.
Not exatctly sure what you mean, but ⇧⌥← and ⇧⌥→ have another meaning (select words) and can't replace ⌥← and ⌥→.

**Comment by dehesselle** [comment]
> I think there no extra work necessary, ⌥← and ⌥→ work fine in the latest versions for me.

This is what a fresh Zim installation looks like, without any pre-existing configuration:
![image](https://user-images.githubusercontent.com/11375494/144141566-5fb51f39-a53f-4469-8d32-902f23145fcd.png)

> Not exatctly sure what you mean, but ⇧⌥← and ⇧⌥→ have another meaning (select words) and can't replace ⌥← and ⌥→.

That is why I was asking and making an example - I'm not familiar with all shortcuts, so the example was just a bad one. The intent is: if Zim has a shortcut for something, better find a replacement than just removing it.

**Comment by kjyv** [comment]
Ok thanks for the info, I was sure I did not unset this shortcut anymore on my current device, but then apparently I was wrong. But anyway my original ticket was about multiple shortcuts, so these should all be specific on macOS if they aren't already. The default GTK3 behavior for the cursor keys is fine now. So yes, open_page_back and _forward should be set to something else, maybe ctrl-⌥-← and -→ (I don't see anything happen on these shortcuts by default in macOS, but please double check).

**Comment by kjyv** [comment]
Another thing I noticed that does not work properly is selecting words with ⇧⌥-↓ etc. but that also seems to be a gtk 3 issue (same behavior in gtk3-demo when installed with homebrew, version 3.24.30). Maybe your dmg can include patches that fix this in Gtk

**Comment by jaap-karssenberg** [comment]
For what it's worth & haven't tested it, but I think you can change key
bindings from the Gtk theme / Gtk CSS configuration. So might not require a
patch, just configuration.

**Comment by kjyv** [comment]
Apart from the non-working shortcuts in my last comment, I have everything configured the way I want it. But I think at least the packaged macOS dmg but ideally also the vanilla version itself should behave properly by default, otherwise I'd consider it a bug. If you consider it the other way round, it would also be weird if you installed zim on Linux or Windows and you would first have to change the macOS shortcuts that are different from your OS default.

**Comment by dehesselle** [comment]
As I want to close this issue, this is what I'm going to do:

- I will adjust the default shortcut for `open_page_back` and `open_page_forward` for the macOS app as discussed. This will only take effect for the macOS app. Doing it correctly within the Python code (I've looked at it; a non-ugly solution is currently beyond my capabilities/resources) is something to be figured out later.

- The (likely upstream) issue about selecting words with ⇧⌥+↓ should be investigated if it can be configured with gtk css as suggested. If not, I can integrate a patch for gtk in my build - but of course this patch needs to be written by someone first. I assume that such a patch doesn't exist already, therefore I consider this out of scope for now.

- I can reproduce the non-working ⌘+g when using the search and wanting to jump to the next search result. I don't have a Linux machine to test it right now, but theoretically it could be a Zim issue. Because invoking the search with ⌘+f shifts focus to the searchbar that pops up in the bottom, deactivating the menu entries (and therefore shortcuts) to perform ⌘+g. If you click into the page after having invoked search, ⌘+g does work as the page having the focus again enables the menu items and their respective shortcuts.
Regardless of what the actual cause turns out to be - Zim or macOS specific - I'd like us to open a dedicated issue for that.

**Comment by jaap-karssenberg** [comment]
Yes, also on linux the "Ctrl-G" keybinding only works when the text is
focussed, not when the searchbar is focussed. Please file this as a
separate ticket and I will have a look.
